### DISCUSSION TITLE
What are the uses of Permission?

### DISCUSSION THREAD

**Discussion by RainyNight9** [tauri-apps/tauri#11706] [Q&A] [answered]
Hello, I would like to make some suggestions.
I feel that there are some problems with the design of Permission? Or is the documentation not clear enough? In short, it is a bit difficult to use now. I can't find a solution quickly and can't develop quickly. I hope it will be more convenient and faster to use and develop quickly!
For example: For example, I am working on a function to create a new window. 

Error:

![image](https://github.com/user-attachments/assets/c6f178b1-efbf-479b-a226-2ebb73f4ed25)

Unhandled Promise Rejection: window.get_all_windows not allowed on window "chat", webview "chat", URL: local
allowed on: [windows: "main", URL: local], [windows: "main", URL: local], [windows: "main", URL: local]
referenced by:...

Use & Config:

```ts
import { getAllWindows, getCurrentWindow } from "@tauri-apps/api/window";
import { WebviewWindow } from "@tauri-apps/api/webviewWindow";

  const closeWin = useCallback(async (label: string) => {
    const targetWindow = await getWin(label);

    if (!targetWindow) {
      return;
    }

    try {
      await targetWindow.close();
    } catch (error) {
    }
  }, []);


  const getWin = useCallback(async (label: string) => {
    return WebviewWindow.getByLabel(label);
  }, []);
```

src-tauri/capabilities/default.json
```json
"permissions": [
    "core:default",
    "core:event:allow-emit",
    "core:event:allow-listen",
    "core:webview:allow-create-webview",
    "core:webview:allow-create-webview-window",
    "core:webview:allow-get-all-webviews",
    "core:webview:allow-webview-close",
    "core:webview:allow-webview-hide",
    "core:webview:allow-webview-show",
    "core:webview:allow-webview-size",
    "core:webview:allow-set-webview-size",
    "core:webview:allow-set-webview-zoom",
    "core:window:default",
    "core:window:allow-center",
    "core:window:allow-close",
    "core:window:allow-hide",
    "core:window:allow-show",
    "core:window:allow-create",
    "core:window:allow-destroy",
    "core:window:allow-start-dragging",
    "core:window:allow-set-size",
    "core:window:allow-get-all-windows",
    "shell:allow-open",
    "http:default",
    "http:allow-fetch",
    "http:allow-fetch-cancel",
    "http:allow-fetch-read-body",
    "http:allow-fetch-send",
    "websocket:default",
    "websocket:allow-connect",
    "websocket:allow-send",
    {
      "identifier": "http:default",
      "allow": [
        {
          "url": "http://localhost:2900"
        }
      ],
      "deny": []
    }
  ]
```

**Comment by FabianLars** [comment, marked as answer]
Your `src-tauri/capabilities/default.json` file should also have a `windows` property by default, doesn't it? I assume it looks like this `"windows": ["main"]` (`main` is the label of your first window unless you specify a custom label). The error is asking you to add the label of your second window to that array because capabilities are opt-in.
`"windows": ["main", "chat"]`

**Reply by RainyNight9** [reply]
Well, thank you very much, the problem has been solved.

However, I still want to ask, why should permissions be separated from tauri.conf.json and placed separately in src-tauri/capabilities/default.json? It feels a bit troublesome to use?
Or is it that the documentation is not clear enough, for example, when you need to call which APIs or other functions, you can't quickly or clearly know which permissions to set?

**Reply by FabianLars** [reply]
> However, I still want to ask, why should permissions be separated from tauri.conf.json and placed separately in src-tauri/capabilities/default.json? It feels a bit troublesome to use?

You can have them in the config, but considering that the capability will potentially get quite large, and because you can define multiple capabilities (files), having them in separate files makes sense (to us).

> Or is it that the documentation is not clear enough, for example, when you need to call which APIs or other functions, you can't quickly or clearly know which permissions to set?

One thing that we may not say explictly in the docs is that the permissions generally match the js function name just in kebab-case and a allow/deny prefix.
Additionally, all the plugin guides have a table of their permissions at the bottom: https://v2.tauri.app/plugin/shell/#permissions

One area that's generally lacking the permissions stuff is the auto-generated [reference](https://v2.tauri.app/reference/javascript/api/), which also includes the IDE's function hover help text. I think this is something we can and should improve.
